oauth2_provider/oauth2_validators.py

```diff
@@ -665,3 +701,174 @@ def validate_refresh_token(self, refresh_token, client, request, *args, **kwargs
+        id_token.update(**claims)
+        # Workaround for oauthlib bug #746
+        # https://github.com/oauthlib/oauthlib/issues/746
+        if "nonce" not in id_token and request.nonce:
+            id_token["nonce"] = request.nonce
+
+        jwt_token = jwt.JWT(
+            header=json.dumps({"alg": request.client.algorithm}, default=str),
+            claims=json.dumps(id_token, default=str),
+        )
+        jwt_token.make_signed_token(request.client.jwk_key)
+        id_token_jwt = jwt_token.serialize()
+        id_token = self._save_id_token(id_token_jwt, request, expiration_time)
+        # this is needed by django rest framework
+        request.access_token = id_token
+        request.id_token = id_token
+        return id_token_jwt
+
+    def validate_jwt_bearer_token(self, token, scopes, request):
+        return self.validate_id_token(token, scopes, request)
+
+    def validate_id_token(self, token, scopes, request):
+        """
+        When users try to access resources, check that provided id_token is valid
+        """
+        if not token:
+            return False
+
+        key = self._get_key_for_token(token)
+        if not key:
+            return False
+        try:
+            jwt_token = jwt.JWT(key=key, jwt=token)
+            id_token = IDToken.objects.get(token=jwt_token.serialize())
+        except (JWException, JWTExpired):
+            return False
+
+        request.client = id_token.application
+        request.user = id_token.user
+        request.scopes = scopes
+        # this is needed by django rest framework
+        request.access_token = id_token
+        return True
+
+    def _get_key_for_token(self, token):
+        """
+        Peek at the unvalidated token to discover who it was issued for
+        and then use that to load that application and its key.
+        """
+        unverified_token = jws.JWS()
+        unverified_token.deserialize(token)
+        claims = json.loads(unverified_token.objects["payload"].decode("utf-8"))
+        if "aud" not in claims:
+            return None
+        application = self._get_client_by_audience(claims["aud"])
+        if application:
+            return application.jwk_key
+
+    def _get_client_by_audience(self, audience):
+        """
+        Load a client by the aud claim in a JWT.
+        aud may be multi-valued, if your provider makes it so.
+        This function is separate to allow further customization.
+        """
+        if isinstance(audience, str):
+            audience = [audience]
+        return Application.objects.filter(client_id__in=audience).first()
+
+    def validate_user_match(self, id_token_hint, scopes, claims, request):
+        # TODO: Fix to validate when necessary acording
+        # https://github.com/idan/oauthlib/blob/master/oauthlib/oauth2/rfc6749/request_validator.py#L556
+        # http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest id_token_hint section
+        return True
```

Comment:
This will have to be implemented later. There is an issue with oauthlib's implementation here. It says:

```
     def validate_user_match(self, id_token_hint, scopes, claims, request):
        """Ensure client supplied user id hint matches session user.

        If the sub claim or id_token_hint is supplied then the session
        user must match the given ID.
```

However, this method is called during the `validate_authorization_request` phase, which gives no opportunity to inject the current session user into the oauthlib `Request`. This means we have no way of comparing the hint to the current session user. :shrug: 

I have a branch with an implementation of this method, but as it stands it would also need some changes to oauthlib.